libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,995 @@
+	}
+
+	auto const sourceNode = findASTNode(dpos.position, dpos.path);
+	if (!sourceNode)
+	{
+		trace(_logPrefix + ": AST node not found for "s + to_string(dpos.position.line) + ":" + to_string(dpos.position.column));
+		// Could not infer AST node from given source location.
+		Json::Value emptyResponse = Json::arrayValue;
+		m_client->reply(_id, emptyResponse);
+		return;
+	}
+
+	vector<SourceLocation> locations;
+
+	bool const handled = matchASTNode(
+		sourceNode,
+		[&](ImportDirective const* importDirective)
+		{
+			auto const& path = *importDirective->annotation().absolutePath;
+
+			auto const i = m_fileReader->sourceCodes().find(path);
+			if (i != m_fileReader->sourceCodes().end())
+				locations.emplace_back(SourceLocation{0, 0, make_shared<CharStream>("", path)});
+			else
+				trace(fmt::format(_logPrefix + ": (importDirective) full path mapping not found for {}: {}\n", importDirective->path(), path));
+		},
+		[&](MemberAccess const* memberAccess)
+		{
+			// For scope members, jump to the naming symbol of the referencing declaration of this member.
+			auto const declaration = memberAccess->annotation().referencedDeclaration;
+			auto const location = declarationPosition(declaration);
+			if (location.has_value())
+				locations.emplace_back(location.value());
+			else
+				trace(_logPrefix + ": declaration not found.");
+		},
+		[&](Identifier const* sourceIdentifier)
+		{
+			// For identifiers, jump to the naming symbol of the definition of this identifier.
+			if (Declaration const* decl = sourceIdentifier->annotation().referencedDeclaration)
+				if (auto location = declarationPosition(decl); location.has_value())
+					locations.emplace_back(move(location.value()));
+
+			for (auto const declaration: sourceIdentifier->annotation().candidateDeclarations)
+				if (auto location = declarationPosition(declaration); location.has_value())
+					locations.emplace_back(move(location.value()));
+		}
+	);
+
+	if (!handled)
+		log(_logPrefix + ": Symbol is not an identifier. "s + typeid(*sourceNode).name());
+
+	Json::Value reply = Json::arrayValue; for (SourceLocation const& location: locations)
+		reply.append(toJson(m_basePath, location));
+
+	m_client->reply(_id, reply);
+}
+
+string LanguageServer::symbolHoverInformation(frontend::ASTNode const* _sourceNode)
+{
+	if (auto const* documented = dynamic_cast<StructurallyDocumented const*>(_sourceNode))
+	{
+		if (documented->documentation())
+			return *documented->documentation()->text();
+	}
+
+	string tooltipText{};
+	matchASTNode(
+		_sourceNode,
+		[&](Identifier const* identifier)
+		{
+			Type const* ty = identifier->annotation().type;
+			tooltipText += ty->toString(false);
+		},
+		[&](MemberAccess const* memberAccess)
+		{
+			if (memberAccess->annotation().type)
+			tooltipText += memberAccess->annotation().type->toString(false);
+		},
+		[&](IdentifierPath const* identifierPath)
+		{
+			Declaration const* decl = identifierPath->annotation().referencedDeclaration;
```

Comment:
Why not use the generic `ASTNode::referencedDeclaration`?

Reply:
Because `ASTNode::referencedDeclaration(...)` expects a `Expression const&` parameter, which I'm not having :)

Reply:
But can't you dynamic_cast it to `Expression const*`? There is really no need to re-write all that code.

Reply:
Done.